contracts/upgradeable_contracts/RewardableValidators.sol

```diff
@@ -0,0 +1,99 @@
+pragma solidity 0.4.24;
+
+import "./BaseBridgeValidators.sol";
+
+
+contract RewardableValidators is BaseBridgeValidators {
+
+    event ValidatorAdded (address indexed validator, address reward);
+    event ValidatorRemoved (address indexed validator);
+
+    function initialize(
+        uint256 _requiredSignatures,
+        address[] _initialValidators,
+        address[] _initialRewards,
+        address _owner
+    )
+    public
+    returns (bool)
+    {
+        require(!isInitialized());
+        require(_owner != address(0));
+        setOwner(_owner);
+        require(_requiredSignatures != 0);
+        require(_initialValidators.length >= _requiredSignatures);
+        require(_initialValidators.length == _initialRewards.length);
+
+        for (uint256 i = 0; i < _initialValidators.length; i++) {
+            require(_initialValidators[i] != address(0) && _initialValidators[i] != F_ADDR);
+            require(_initialRewards[i] != address(0));
+            require(!isValidator(_initialValidators[i]));
+
+            if (i == 0) {
+                setValidator(F_ADDR, _initialValidators[i]);
+                if (_initialValidators.length == 1) {
+                    setValidator(_initialValidators[i], F_ADDR);
+                }
+            } else if (i == _initialValidators.length - 1) {
+                setValidator(_initialValidators[i - 1], _initialValidators[i]);
+                setValidator(_initialValidators[i], F_ADDR);
+            } else {
+                setValidator(_initialValidators[i - 1], _initialValidators[i]);
+            }
+
+            setValidatorCount(validatorCount().add(1));
+            setValidatorRewardAddress(_initialValidators[i], _initialRewards[i]);
+            emit ValidatorAdded(_initialValidators[i], _initialRewards[i]);
+        }
+
+        uintStorage[keccak256(abi.encodePacked("requiredSignatures"))] = _requiredSignatures;
+        uintStorage[keccak256("deployedAtBlock")] = block.number;
+        setInitialize(true);
+        emit RequiredSignaturesChanged(_requiredSignatures);
+
+        return isInitialized();
+    }
+
+    function addValidator(address _validator, address _reward) external onlyOwner {
+        require(_validator != address(0) && _validator != F_ADDR);
+        require(_reward != address(0));
+        require(!isValidator(_validator));
+
+        address firstValidator = getNextValidator(F_ADDR);
+        setValidator(_validator, firstValidator);
+        setValidatorRewardAddress(_validator, _reward);
```

Comment:
So, the difference in this method between two contracts is actually: 

```solidity
setValidatorRewardAddress(_validator, _reward);
emit ValidatorAdded(_validator, _reward);
```

Could we have the common functionality in the basic contract and invoke it in the child contract?
For example, `BridgeValidators` will have
```solidity
function addValidator(address _validator) external onlyOwner {
    _addValidator(_validator);
    emit ValidatorAdded(_validator);
}
```
and  `RewardableValidators` will have
```solidity
function addValidator(address _validator) external onlyOwner {
    _addValidator(_validator);
    setValidatorRewardAddress(_validator, _reward);
    emit ValidatorAdded(_validator, _reward);
}
```

Similar is for the `removeValidator` functionality.

Reply:
Thanks for suggestion! Fixed on https://github.com/poanetwork/poa-bridge-contracts/pull/130/commits/b74360256415b3fe2f5aca4a9e0a486655643caf